lib/db/instance.go

```diff
@@ -372,81 +522,168 @@ func (db *instance) withNeed(folder, device []byte, truncate bool, fn Iterator)
 			l.Debugf("need folder=%q device=%v name=%q have=%v invalid=%v haveV=%v globalV=%v globalDev=%v", folder, devID, name, have, haveFV.Invalid, haveFV.Version, needVersion, needDevice)
 
 			if !fn(gf) {
-				return
+				return nil
 			}
 
 			// This file is handled, no need to look further in the version list
 			break
 		}
 	}
+	return dbi.Error()
```

Comment:
Same question as @imsodin, the relation between .Next() and Error() is not clear.

We should declare how it works in the interface, and make sure leveldb guarantees that behaviour.